### PR TITLE
Line numbers for better error handling (WIP)

### PR DESCRIPTION
**Pull request by kunal-mohta**
This presents a proposal for solving #40. 
I have used the method to solve #30. You can check the method by trying to reproduce the mentioned there.
If this looks fine, I can try to implement this on all the errors thrown in the code.

### CODE REVIEW

**Comment by stebanos** [comment]
I'm not a fan of having the line number in Lexer. Since Parser manages the progress of the Lexer, it could catch the errors thrown in Lexer (and itself) and rethrow them with the line number added. That way, errors in Lexer don't need to be rewritten. Not for that reason anyway. Also for added consistency with the code we have already the error message should start with ```Line ${lineno}:```

**Comment by kunal-mohta** [comment]
@stebanos This looks better?

**Comment by stebanos** [comment]
Hey Kunal, yes this is better. Will you also add in the error re-throwing?

**Comment by kunal-mohta** [comment]
Sure!

**Comment by kunal-mohta** [comment]
There are still errors thrown in the `Interpreter` class that need line numbers.
I guess that would need line numbers to be present in the `phraseBook` as well?

**Comment by stebanos** [comment]
Hehe, I think you misunderstood my last request. You added ```Line ${this.lineno}```in front of each error. What I meant was something like this:

in the PhraseParser:
```
    parse() {
        try {
            const node = this.phrase();
            if (this.currentToken.type !== EOF) {
                this.error(EOF);
            }
            return node;
        } catch (e) {
            throw new Error(`Line ${this.lineno}: ${e.message}`);
        }
    }
```

A similar thing could be done in the Interpreter class if the phrase has access to its line number.

**Comment by kunal-mohta** [comment]
Oh! :man_facepalming: 
I am sorry.. will do that. :sweat_smile: 
For giving `Interpreter` class the access to line numbers I would need to add line numbers to the `phraseBook` right? Or is there a better way to do it?

**Comment by stebanos** [comment]
The Interpreter can access the phrase through ```this.phrase```, in the screenshot you showed earlier ```lineno``` was included in the phrase, so I suppose ```this.phrase.lineno```would just work?

**Comment by kunal-mohta** [comment]
Yes, that is what I plan to do. :+1:

**Comment by kunal-mohta** [comment]
Sorry about the commit message :sweat:

**Comment by stebanos** [comment]
Thanks, well done!

File: js/core.js [hunk 1]
```diff
@@ -434,7 +434,7 @@ class PhraseLexer extends Lexer {
             } else if (this.checkEscapeChar('}')) {
                 result += '}';
                 this.advance();
-                this.advance();
+                this.advance();line
```
Comment (fdb) [inline js/core.js:437]: I know this is now merged, but I don't understand this code. What's the line at the end? Just a typo?

Comment (kunal-mohta) [reply js/core.js:437]: I think that is what it is :sweat: 
I am sorry, let me just change it.

Comment (kunal-mohta) [reply js/core.js:437]: @fdb I have made the correction in #43

Comment (fdb) [reply js/core.js:437]: 👍

**Pull request merged**